src/base/bittorrent/tracker.cpp

```diff
@@ -216,50 +216,62 @@ void Tracker::respondToAnnounceRequest()
     // 7. TODO: support "compact" extension
 
     // Done parsing, now let's reply
-    if (m_torrents.contains(annonceReq.infoHash)) {
-        if (annonceReq.event == "stopped") {
-            qDebug("Tracker: Peer stopped downloading, deleting it from the list");
-            m_torrents[annonceReq.infoHash].remove(annonceReq.peer.uid());
-            return;
-        }
+    if (annonceReq.event == "stopped") {
+        unregisterPeer(annonceReq);
     }
     else {
+        registerPeer(annonceReq);
+        replyWithPeerList(annonceReq);
+    }
+}
+
+void Tracker::registerPeer(const TrackerAnnounceRequest &annonceReq)
+{
+    if (annonceReq.peer.port == 0) return;
+
+    if (!m_torrents.contains(annonceReq.infoHash)) {
         // Unknown torrent
         if (m_torrents.size() == MAX_TORRENTS) {
             // Reached max size, remove a random torrent
             m_torrents.erase(m_torrents.begin());
         }
     }
+
     // Register the user
-    PeerList peers = m_torrents.value(annonceReq.infoHash);
-    if (peers.size() == MAX_PEERS_PER_TORRENT) {
-        // Too many peers, remove a random one
-        peers.erase(peers.begin());
+    PeerList &peers = m_torrents[annonceReq.infoHash];
+    if (!peers.contains(annonceReq.peer.uid())) {
+        // Unknown peer
+        if (peers.size() == MAX_PEERS_PER_TORRENT) {
+            // Too many peers, remove a random one
+            peers.erase(peers.begin());
+        }
     }
     peers[annonceReq.peer.uid()] = annonceReq.peer;
-    m_torrents[annonceReq.infoHash] = peers;
+}
+
+void Tracker::unregisterPeer(const TrackerAnnounceRequest &annonceReq)
+{
+    if (annonceReq.peer.port == 0) return;
```

Comment:
Do you need this check? I don't see the full picture, just double confirming with you.

Reply:
Just optimization. Don't lookup for peers with port=0 since we register no such peers.